File: src/common/gladys_common/logging.py
"""Structured logging for GLADyS services.

Provides consistent, debuggable logging across all services with:
- Structured output (JSON or human-readable)
- Trace ID propagation for request correlation
- File and console output
- Configurable log levels

Usage:
    from gladys_common import setup_logging, get_logger, bind_trace_id

    # At service startup
    setup_logging("memory-python")

    # In request handlers
    logger = get_logger()
    bind_trace_id(trace_id)  # From gRPC metadata or generate new
    logger.info("Processing request", request_id=req.id)

Configuration via environment variables:
    LOG_LEVEL: DEBUG, INFO, WARN, ERROR (default: INFO)
    LOG_FORMAT: human, json (default: human)
    LOG_FILE: Path to log file (optional)
    LOG_FILE_LEVEL: Level for file output (default: same as LOG_LEVEL)
"""

import logging
import os
import secrets
import sys
from logging.handlers import RotatingFileHandler
from pathlib import Path
from typing import Any

import structlog

# Constants
TRACE_ID_HEADER = "x-gladys-trace-id"
DEFAULT_LOG_LEVEL = "INFO"
DEFAULT_LOG_FORMAT = "human"
LOG_FILE_MAX_BYTES = 10 * 1024 * 1024  # 10MB
LOG_FILE_BACKUP_COUNT = 5


def generate_trace_id() -> str:
    """Generate a new trace ID (12 hex characters)."""
    return secrets.token_hex(6)


def _get_log_level(level_str: str) -> int:
    """Convert string log level to logging constant."""
    levels = {
        "DEBUG": logging.DEBUG,
        "INFO": logging.INFO,
        "WARN": logging.WARNING,
        "WARNING": logging.WARNING,
        "ERROR": logging.ERROR,
    }
    return levels.get(level_str.upper(), logging.INFO)


def _create_file_handler(log_file: str, level: int) -> RotatingFileHandler:
    """Create a rotating file handler."""
    # Ensure log directory exists
    log_path = Path(log_file)
    log_path.parent.mkdir(parents=True, exist_ok=True)

    handler = RotatingFileHandler(
        log_file,
        maxBytes=LOG_FILE_MAX_BYTES,
        backupCount=LOG_FILE_BACKUP_COUNT,
        encoding="utf-8",
    )
    handler.setLevel(level)
    return handler


def setup_logging(service_name: str) -> None:
    """Configure logging for a GLADyS service.

    Args:
        service_name: Name of the service (e.g., "memory-python", "orchestrator")
    """
    level_str = os.environ.get("LOG_LEVEL", DEFAULT_LOG_LEVEL)
    log_format = os.environ.get("LOG_FORMAT", DEFAULT_LOG_FORMAT)
    log_file = os.environ.get("LOG_FILE")
    file_level_str = os.environ.get("LOG_FILE_LEVEL", level_str)

    level = _get_log_level(level_str)
    file_level = _get_log_level(file_level_str)

    # Configure standard logging (structlog will use this)
    root_logger = logging.getLogger()
    root_logger.setLevel(min(level, file_level))

    # Remove existing handlers
    root_logger.handlers.clear()

    # Console handler
    console_handler = logging.StreamHandler(sys.stderr)
    console_handler.setLevel(level)
    root_logger.addHandler(console_handler)

    # File handler (if configured)
    if log_file:
        file_handler = _create_file_handler(log_file, file_level)
        root_logger.addHandler(file_handler)

    # Configure structlog to use stdlib logging (so file handlers work)
    # Pre-chain processors run before passing to stdlib
    pre_chain = [
        structlog.contextvars.merge_contextvars,
        structlog.processors.add_log_level,
        structlog.processors.TimeStamper(fmt="iso"),
        structlog.stdlib.ExtraAdder(),
    ]

    if log_format == "json":
        # JSON formatter for stdlib handler
        formatter = structlog.stdlib.ProcessorFormatter(
            processors=[
                structlog.stdlib.ProcessorFormatter.remove_processors_meta,
                structlog.processors.JSONRenderer(),
            ],
            foreign_pre_chain=pre_chain,
        )
    else:
        # Human-readable formatter for stdlib handler
        formatter = structlog.stdlib.ProcessorFormatter(
            processors=[
                structlog.stdlib.ProcessorFormatter.remove_processors_meta,
                structlog.dev.ConsoleRenderer(colors=True),
            ],
            foreign_pre_chain=pre_chain,
        )

    # Apply formatter to all handlers
    for handler in root_logger.handlers:
        handler.setFormatter(formatter)

    # Configure structlog to use stdlib
    structlog.configure(
        processors=pre_chain + [
            structlog.stdlib.ProcessorFormatter.wrap_for_formatter,
        ],
        wrapper_class=structlog.stdlib.BoundLogger,
        context_class=dict,
        logger_factory=structlog.stdlib.LoggerFactory(),
        cache_logger_on_first_use=True,
    )

    # Bind service name to all logs
    structlog.contextvars.clear_contextvars()
    structlog.contextvars.bind_contextvars(service=service_name)


def get_logger(name: str | None = None) -> structlog.stdlib.BoundLogger:
    """Get a configured logger.

    Args:
        name: Optional logger name. If not provided, uses root logger.

    Returns:
        A structlog BoundLogger with service context already bound.
    """
    return structlog.get_logger(name)


def bind_trace_id(trace_id: str) -> None:
    """Bind a trace ID to the current context.

    Call this at the start of each request to enable request correlation
    across service boundaries.

    Args:
        trace_id: The trace ID (usually from gRPC metadata or generated)
    """
    structlog.contextvars.bind_contextvars(trace_id=trace_id)


def unbind_trace_id() -> None:
    """Remove the trace ID from the current context.

    Call this at the end of a request if the context will be reused.
    """
    structlog.contextvars.unbind_contextvars("trace_id")


def extract_trace_id_from_metadata(metadata: Any) -> str | None:
    """Extract trace ID from gRPC metadata.

    Args:
        metadata: gRPC invocation metadata (dict-like)

    Returns:
        The trace ID if present, None otherwise.
    """
    if hasattr(metadata, "get"):
        return metadata.get(TRACE_ID_HEADER)
    # Handle list of tuples format
    if isinstance(metadata, (list, tuple)):
        for key, value in metadata:
            if key == TRACE_ID_HEADER:
                return value
    return None


def get_or_create_trace_id(metadata: Any) -> str:
    """Get trace ID from metadata or generate a new one.

    Args:
        metadata: gRPC invocation metadata

    Returns:
        Existing trace ID from metadata, or a newly generated one.
    """
    trace_id = extract_trace_id_from_metadata(metadata)
    if not trace_id:
        trace_id = generate_trace_id()
    return trace_id

File: src/services/salience/src/logging.rs
//! Structured logging for GLADyS Rust services.
//!
//! Provides consistent, debuggable logging matching the Python services with:
//! - Structured output (JSON or human-readable)
//! - Trace ID propagation for request correlation
//! - File and console output
//! - Configurable log levels
//!
//! Configuration via environment variables:
//!   LOG_LEVEL: trace, debug, info, warn, error (default: info)
//!   LOG_FORMAT: human, json (default: human)
//!   LOG_FILE: Path to log file (optional)

use std::env;
use std::path::Path;
use tracing_appender::non_blocking::WorkerGuard;
use tracing_subscriber::{
    fmt,
    layer::SubscriberExt,
    util::SubscriberInitExt,
    EnvFilter,
};

/// Guard for non-blocking file writer. Must be held for the lifetime of the application.
pub struct LogGuard {
    _file_guard: Option<WorkerGuard>,
}

/// Initialize logging for a GLADyS Rust service.
///
/// Returns a guard that must be held for the application lifetime to ensure
/// logs are flushed to file.
pub fn setup_logging(service_name: &str) -> LogGuard {
    let log_level = env::var("LOG_LEVEL").unwrap_or_else(|_| "info".to_string());
    let log_format = env::var("LOG_FORMAT").unwrap_or_else(|_| "human".to_string());
    let log_file = env::var("LOG_FILE").ok();

    // Build env filter - use LOG_LEVEL if RUST_LOG isn't set
    let filter = env::var("RUST_LOG").unwrap_or_else(|_| {
        format!("{}={}", env!("CARGO_PKG_NAME").replace("-", "_"), log_level)
    });
    let env_filter = EnvFilter::try_new(&filter).unwrap_or_else(|_| EnvFilter::new("info"));

    // Build subscriber based on format and file options
    // Using separate paths to avoid type mismatches between JSON and human formats
    let file_guard = match (log_format.as_str(), log_file) {
        ("json", Some(path)) => {
            let path = Path::new(&path);
            let dir = path.parent().unwrap_or(Path::new("."));
            let filename = path.file_name().and_then(|n| n.to_str()).unwrap_or("service.log");
            let file_appender = tracing_appender::rolling::never(dir, filename);
            let (non_blocking, guard) = tracing_appender::non_blocking(file_appender);

            tracing_subscriber::registry()
                .with(env_filter)
                .with(fmt::layer().json().with_target(false).with_file(false).with_line_number(false))
                .with(fmt::layer().json().with_writer(non_blocking).with_target(false).with_file(false).with_line_number(false))
                .init();
            Some(guard)
        }
        ("json", None) => {
            tracing_subscriber::registry()
                .with(env_filter)
                .with(fmt::layer().json().with_target(false).with_file(false).with_line_number(false))
                .init();
            None
        }
        (_, Some(path)) => {
            let path = Path::new(&path);
            let dir = path.parent().unwrap_or(Path::new("."));
            let filename = path.file_name().and_then(|n| n.to_str()).unwrap_or("service.log");
            let file_appender = tracing_appender::rolling::never(dir, filename);
            let (non_blocking, guard) = tracing_appender::non_blocking(file_appender);

            tracing_subscriber::registry()
                .with(env_filter)
                .with(fmt::layer().with_ansi(true).with_target(false).with_file(false).with_line_number(false))
                .with(fmt::layer().with_writer(non_blocking).with_ansi(false).with_target(false).with_file(false).with_line_number(false))
                .init();
            Some(guard)
        }
        (_, None) => {
            tracing_subscriber::registry()
                .with(env_filter)
                .with(fmt::layer().with_ansi(true).with_target(false).with_file(false).with_line_number(false))
                .init();
            None
        }
    };

    // Log startup info
    tracing::info!(service = service_name, "Logging initialized");

    LogGuard {
        _file_guard: file_guard,
    }
}

/// Generate a new trace ID (12 hex characters).
pub fn generate_trace_id() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos();
    format!("{:012x}", now & 0xffffffffffff)
}

/// Header name for trace ID in gRPC metadata.
pub const TRACE_ID_HEADER: &str = "x-gladys-trace-id";

/// Extract trace ID from gRPC metadata.
pub fn extract_trace_id<T>(request: &tonic::Request<T>) -> Option<String> {
    request
        .metadata()
        .get(TRACE_ID_HEADER)
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string())
}

/// Get trace ID from request or generate a new one.
pub fn get_or_create_trace_id<T>(request: &tonic::Request<T>) -> String {
    extract_trace_id(request).unwrap_or_else(generate_trace_id)
}
